libraries/AP_Radio/AP_Radio_cypress.cpp

```diff
@@ -0,0 +1,1748 @@
+#define CYRF_TXBERR_IRQEN (1<<2)
+#define CYRF_TXB0_IRQEN   (1<<3)
+#define CYRF_TXB8_IRQEN   (1<<4)
+#define CYRF_TXB15_IRQEN  (1<<5)
+#define CYRF_TX_CLR       (1<<6)
+#define CYRF_TX_GO        (1<<7)
+
+// CYRF_RX_CTRL
+#define CYRF_RXE_IRQEN    (1<<0)
+#define CYRF_RXC_IRQEN    (1<<1)
+#define CYRF_RXBERR_IRQEN (1<<2)
+#define CYRF_RXB1_IRQEN   (1<<3)
+#define CYRF_RXB8_IRQEN   (1<<4)
+#define CYRF_RXB16_IRQEN  (1<<5)
+#define CYRF_RSVD         (1<<6)
+#define CYRF_RX_GO        (1<<7)
+
+// CYRF_RX_OVERRIDE
+#define CYRF_ACE          (1<<1)
+#define CYRF_DIS_RXCRC    (1<<2)
+#define CYRF_DIS_CRC0     (1<<3)
+#define CYRF_FRC_RXDR     (1<<4)
+#define CYRF_MAN_RXACK    (1<<5)
+#define CYRF_RXTX_DLY     (1<<6)
+#define CYRF_ACK_RX       (1<<7)
+
+// CYRF_TX_OVERRIDE
+#define CYRF_TX_INV       (1<<0)
+#define CYRF_DIS_TXCRC    (1<<2)
+#define CYRF_OVRD_ACK     (1<<3)
+#define CYRF_MAN_TXACK    (1<<4)
+#define CYRF_FRC_PRE      (1<<6)
+#define CYRF_ACK_TX       (1<<7)
+
+// CYRF_RX_CFG
+#define CYRF_VLD_EN       (1<<0)
+#define CYRF_RXOW_EN      (1<<1)
+#define CYRF_FAST_TURN_EN (1<<3)
+#define CYRF_HILO         (1<<4)
+#define CYRF_ATT          (1<<5)
+#define CYRF_LNA          (1<<6)
+#define CYRF_AGC_EN       (1<<7)
+
+// CYRF_TX_CFG
+enum {
+    CYRF_PA_M35      = 0x0,
+    CYRF_PA_M30      = 0x1,
+    CYRF_PA_M24      = 0x2,
+    CYRF_PA_M18      = 0x3,
+    CYRF_PA_M13      = 0x4,
+    CYRF_PA_M5       = 0x5,
+    CYRF_PA_0        = 0x6,
+    CYRF_PA_4        = 0x7,
+};
+enum {
+    CYRF_DATA_MODE_GFSK   = (0x0 <<3),
+    CYRF_DATA_MODE_8DR    = (0x1 <<3),
+    CYRF_DATA_MODE_DDR    = (0x2 <<3),
+    CYRF_DATA_MODE_SDR    = (0x3 <<3),
+};
+#define CYRF_DATA_CODE_LENGTH    (1<<5)
+
+
+#define FLAG_WRITE      0x80
+#define FLAG_AUTO_INC   0x40
+
+#define DSM_MAX_CHANNEL 0x4F
+
+#define DSM_SCAN_MIN_CH 8
+#define DSM_SCAN_MID_CH 40
+#define DSM_SCAN_MAX_CH 70
+
+#define FCC_SUPPORT_CW_MODE 0
+
+#define AUTOBIND_CHANNEL 12
+
+// object instance for trampoline
+AP_Radio_cypress *AP_Radio_cypress::radio_instance;
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+thread_t *AP_Radio_cypress::_irq_handler_ctx;
+#endif
+/*
+  constructor
+ */
+AP_Radio_cypress::AP_Radio_cypress(AP_Radio &_radio) :
+    AP_Radio_backend(_radio)
+{
+    // link to instance for irq_trampoline
+    radio_instance = this;
+}
+
+/*
+  initialise radio
+ */
+bool AP_Radio_cypress::init(void)
+{
+    dev = hal.spi->get_device(CYRF_SPI_DEVICE);
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    if(_irq_handler_ctx != nullptr) {
+        AP_HAL::panic("AP_Radio_cypress: double instantiation of irq_handler\n");
+    }
+    chVTObjectInit(&timeout_vt);
+    _irq_handler_ctx = chThdCreateStatic(_irq_handler_wa,
+                     sizeof(_irq_handler_wa),
+                     TIMEOUT_PRIORITY,        /* Initial priority.    */
+                     irq_handler_thd,  /* Thread function.     */
+                     NULL);                     /* Thread parameter.    */
```

Comment:
I thought the point of the 'hal' was to hid this sort of OS specific thing from the ArduPilot code.   Maybe this chibios "virtual timer" chVTObjectInit  and 'Thread'  chThdCreateStatic  activities should be put in a hal->timer() or hal->thread() or something less platform specific?

Reply:
yes indeed - and Sid has some patches to add APIs for this